plonkish_backend/src/pcs/multilinear/mercury.rs

```diff
@@ -0,0 +1,452 @@
+        transcript::{TranscriptRead, TranscriptWrite},
+        DeserializeOwned, Itertools, Serialize,
+    },
+    Error,
+};
+use halo2_curves::{ff::PrimeField, CurveAffine};
+use rand::RngCore;
+use std::marker::PhantomData;
+
+#[derive(Clone, Debug)]
+pub struct Mercury<Pcs>(PhantomData<Pcs>);
+
+impl<M> PolynomialCommitmentScheme<M::Scalar> for Mercury<UnivariateKzg<M>>
+where
+    M: MultiMillerLoop,
+    M::Scalar: Serialize + DeserializeOwned,
+    M::G1Affine: Serialize + DeserializeOwned + CurveAffine<ScalarExt = M::Scalar>,
+    M::G2Affine: Serialize + DeserializeOwned + CurveAffine<ScalarExt = M::Scalar>,
+{
+    type Param = <UnivariateKzg<M> as PolynomialCommitmentScheme<M::Scalar>>::Param;
+    type ProverParam = <UnivariateKzg<M> as PolynomialCommitmentScheme<M::Scalar>>::ProverParam;
+    type VerifierParam = <UnivariateKzg<M> as PolynomialCommitmentScheme<M::Scalar>>::VerifierParam;
+    type Polynomial = MultilinearPolynomial<M::Scalar>;
+    type Commitment = <UnivariateKzg<M> as PolynomialCommitmentScheme<M::Scalar>>::Commitment;
+    type CommitmentChunk =
+        <UnivariateKzg<M> as PolynomialCommitmentScheme<M::Scalar>>::CommitmentChunk;
+
+    fn setup(poly_size: usize, batch_size: usize, rng: impl RngCore) -> Result<Self::Param, Error> {
+        UnivariateKzg::<M>::setup(poly_size, batch_size, rng)
+    }
+
+    fn trim(
+        param: &Self::Param,
+        poly_size: usize,
+        batch_size: usize,
+    ) -> Result<(Self::ProverParam, Self::VerifierParam), Error> {
+        UnivariateKzg::<M>::trim(param, poly_size, batch_size)
+    }
+
+    fn commit(pp: &Self::ProverParam, poly: &Self::Polynomial) -> Result<Self::Commitment, Error> {
+        if pp.degree() + 1 < poly.evals().len() {
+            let got = poly.evals().len() - 1;
+            return Err(err_too_large_deree("commit", pp.degree(), got));
+        }
+
+        Ok(UnivariateKzg::commit_monomial(pp, poly.evals()))
+    }
+
+    fn batch_commit<'a>(
+        pp: &Self::ProverParam,
+        polys: impl IntoIterator<Item = &'a Self::Polynomial>,
+    ) -> Result<Vec<Self::Commitment>, Error> {
+        polys
+            .into_iter()
+            .map(|poly| Self::commit(pp, poly))
+            .collect()
+    }
+
+    fn open(
+        pp: &Self::ProverParam,
+        poly: &Self::Polynomial,
+        comm: &Self::Commitment,
+        point: &Point<M::Scalar, Self::Polynomial>,
+        eval: &M::Scalar,
+        transcript: &mut impl TranscriptWrite<Self::CommitmentChunk, M::Scalar>,
+    ) -> Result<(), Error> {
+        let num_vars = point.len();
+        if pp.degree() + 1 < poly.evals().len() {
+            let got = poly.evals().len() - 1;
+            return Err(err_too_large_deree("open", pp.degree(), got));
+        }
+
+        if cfg!(feature = "sanity-check") {
+            assert_eq!(Self::commit(pp, poly).unwrap().0, comm.0);
+            assert_eq!(poly.evaluate(point), *eval);
+        }
+
+        let t = (num_vars >> 1) + 1;
```

Comment:
This seems to add some overhead when `num_vars` is divisible by 2. 

```suggestion
        let t = num_vars.div_ceil(2);
```